signals/signal_exporter.py

```diff
@@ -0,0 +1,383 @@
+        try:
+            data = json.loads(response["data"])
+        except ValueError:
+            return None
+        if "tool_id" not in data or "publisher_id" not in data or "event" not in data:
+            print("Malformed response data found")
+            return None
+        return data
+
+    def _fetch_responders(self) -> None:
+        """
+        Start initialization response listener. Add tool_ids from proper
+        responses to the subscriber list.
+        """
+        subscriber = self.redis.pubsub(ignore_subscribe_messages=True)
+
+        def _init_handler(item) -> None:
+            data = self._get_data_dict(item)
+            if (
+                data
+                and data["event"] == "initialization"
+                and data["publisher_id"] == self.pub_id
+            ):
+                self.subs.append(data["tool_id"])
+
+        subscriber.subscribe(**{"event-signal-response": _init_handler})
+        self.init_listener = subscriber.run_in_thread()
+
+    def _check_subs(self, event: str) -> Tuple[Any, list]:
+        """
+        Listen for responses from all registered subscribers. Return
+        listener, as well as value based on responders' RAS codes.
+        """
+        if not self.subs:
+            return None, [0]
+
+        to_check = set(self.subs)
+        subscriber = self.redis.pubsub(ignore_subscribe_messages=True)
+        result_box = [0]
+
+        def _sub_handler(item: Dict) -> None:
+            data = self._get_data_dict(item)
+            if data and data["publisher_id"] == self.pub_id and data["event"] == event:
+                if "ras" in data:
+                    to_check.remove(data["tool_id"])
+                    if data["ras"] != 1:
+                        print(
+                            f"WARNING: Tool '{data['tool_id']}' returned bad response for event '{event}', ras: {data['ras']}"
+                        )
+                        result_box[0] = 1
+            if not to_check:
+                listener.stop()
+
+        subscriber.subscribe(**{"event-signal-response": _sub_handler})
+        listener = subscriber.run_in_thread()
+        return listener, result_box
+
+    def _valid_str_list(self, names: list) -> bool:
+        """
+        Return true if input is a non-empty list of strings. Otherwise
+        return false.
+        """
+        return (
+            bool(names)
+            and isinstance(names, list)
+            and all(isinstance(event, str) for event in names)
+        )
+
+    def publish_signal(
+        self, event: str, sample: int = -1, tag: str = None, metadata: Dict = None
+    ) -> int:
+        """
+        Publish a legal event signal. Includes additional options to specify sample_no,
+        a tag, and any other additional metadata. Will then wait for responses from
+        subscribed responders (if any). Returns one of the below result codes based on 
+        signal publish/response success.
+
+        RESULT CODES:
```

Comment:
Can we add global constants to make this more readable? So up top define:

```python
ALL_SUBS_WELL = 0
SUB_FAILED = 1
MISSING_RESPONSE = 2
...
```

Might make the code in this function easier to read right off the bat